### ISSUE TITLE
Build on darwin for Docker container

### ISSUE THREAD

**Issue by ojkelly** [bazelbuild/rules_go#354]
Hi,

I want to use Bazel to build a go binary on MacOS for a docker container.

I have a working build, but it's built for darwin and fails inside the container.

I've read through all the talk in #98 #145 #70, and it seems like this is blocked by Bazel working out cross compilation. But that sounds like it wont be resolved until mid to late this year.

The destination of the build will always be linux/amd64.

I've tried dazel (Bazel in docker) but its slow and a bit buggy.
I tried using #98, but it's currently broken.

Is there any way this could work?
Does anyone have a solution for this yet?

Thanks for your help!

**Comment by jayconrod** [comment]
Unfortunately, I don't have good news for you. You've accurately described the current state of the world. There are three major problems to solve.

1. We need to set `GOOS` and `GOARCH`, both for host and target configurations. Bazel platforms and toolchains should provide a nice way to do this. I've experimented with having Bazel read them from the environment and with setting them using --define, but these are pretty ugly approaches, and we don't want to support that long-term.
1. We need to be able to build the standard library. In `go_repositories`, we download an archive that contains binaries for the Go toolchain and standard libraries for the host platform only, so we'll need to compile the standard library for the target platform. This shouldn't be too difficult to solve, but we still need to know `GOOS` and `GOARCH`.
1. We need to make this work for Cgo and external linkers. Bazel supports cross-compiling C and C++ code with --crosstool_top, but configuring crosstool is painful. C and C++ rules may be covered by the new platforms and toolchains in Bazel eventually, but not as soon as Skylark rules (pure Go).

**Comment by ojkelly** [comment]
Thanks @jayconrod that's what I suspected. Even with a single build target it's not doable at the moment.

I'll give dazel another go.

**Comment by macrael** [comment]
Take a look at https://jpetazzo.github.io/2015/09/03/do-not-use-docker-in-docker-for-ci/

Right now I'm working with bazel in docker directly instead of using dazel to proxy commands across. I'd be curious if dazel makes things significantly easier.

**Comment by ojkelly** [comment]
I ended up setting up the entire docker VM to run in Vagrant, and using Bazel inside there. Eliminates the need to run a local docker repository, and means the go binaries are built on the same platform as the destination. It's working pretty well.

**Comment by ashi009** [comment]
Hey @jayconrod, any process since your last comment?

**Comment by jayconrod** [comment]
@ashi009 Quite a bit of progress, yes. As of rules_go [0.7.0](https://github.com/bazelbuild/rules_go/releases/tag/0.7.0), you can cross compile by adding a platform flag to the command line like this:

```
bazel build --experimental_platforms=@io_bazel_rules_go//go/toolchain:linux_amd64 //:my_binary
```

Just adjust the toolchain target for your GOOS / GOARCH. Anything that the Go SDK supports should work.

There are still some known issues:

* Cross compilation only works for pure Go targets. There are some issues with cgo that we're working out.
* Platforms aren't connected with the --cpu flag, which Gazelle uses to filter platform-specific sources and deps. We'll fix this after Bazel 0.8 is released.
* There is also an issue with Windows executable extensions that will be fixed after Bazel 0.8.

None of this is specific to Docker -- you would just use a normal Bazel installation for this. [rules_docker](https://github.com/bazelbuild/rules_docker) may be able to help with the rest.

cc @ianthehat

**Comment by ashi009** [comment]
Thanks for the explanation!
Jay Conrod <notifications@github.com>于2017年11月22日 周三下午11:48写道：
